### DISCUSSION TITLE
Feature Requests/Ideas

### DISCUSSION THREAD

**Discussion by rcarriga** [nvim-neotest/neotest#20] [Ideas]
Some ideas that may or may not be implemented in future:

- File watching: Auto run tests when relevant code has updated
- Code coverage: Display code coverage using signs/virtual text/highlighting
- Configurable discovery: Tweak test discovery for performance https://github.com/nvim-neotest/neotest/issues/13
- Per-project configuration: Change adapters, discovery, potentially other config depending on project directory https://github.com/nvim-neotest/neotest/issues/13

**Comment by olimorris** [comment]
File watching - I love the idea of having the summary split open and being able to see how my whole test suite is affected by my changes.

Code coverage - This would also be great to have and I use [nvim-coverage ](https://github.com/andythigpen/nvim-coverage) now to pretty awesome effect:

<img width="845" alt="Screen Shot 2022-06-13 at 08 42 28@2x" src="https://user-images.githubusercontent.com/9512444/173304277-8c28bbee-7b8e-4813-8cc9-ba78586a69a3.png">

Ideas from my side:
- Documentation: Would love there to be a wiki or guide on how to create adapters (I'd be happy to contribute). Having completed the RSpec one, most of it was really straight forward but the challenging part was collecting results and reaching out to yourself solved a big headache. Would be great to highlight and provide some solutions to those problems to help get the adapter community growing.
- Adapter tests: I tried to write some tests for the RSpec adapter but hit some weird error. Would be great to ensure that adapters work with specific versions of neotest (assuming you're going to move to some versioning mechanism at some point) and that we can easily run them against specific edge cases. For the RSpec adapter I'm envisaging users coming to me with edge cases that I need to manually test against. Danger is that I then break other cases in trying to account for those edge cases. Some form of adapter testing would be amazing.

**Reply by rcarriga** [reply]
Thanks for the the pointers on the docs! Will get around to making a fleshed out wiki entry to make things clearer.

What issue were you having with running tests? Fairly used to writing plenary tests at this point so should be able to help out. One obvious one is to make sure for any async functions to use the async `it` wrapper. Examples are in the `neotest` repo

Also neotest now follow semantic versioning so it'll be clear when there is a breaking change. Hopefully one day plugin managers can handle more than just equality versions so users can stay up to date until a breaking change.

**Reply by olimorris** [reply]
Regarding the tests, we've actually got some nice ones running over at https://github.com/olimorris/neotest-rspec. We were running into a "c-boundary" error which was resolved by making the required tests, async. One thing I've yet to do is write some full integration tests with neotest and the adapter working.

Also, regarding the guide on creating adapters...I'd be happy to make a first attempt.

**Reply by rcarriga** [reply]
It'd be great to have someone take a stab at docs. If you'd like to, I would appreciate it! Of course also happy to fill in the gaps if you find some things unclear

**Comment by gugahoi** [comment]
Just curious if there are any updates regarding `File watching: Auto run tests when relevant code has updated`?

This would be a great addition! Alternatively, if anyone is willing to share, how are others workflows?

**Reply by stevearc** [reply]
The ideal would be something sophisticated enough to determine exactly which files have changed, exactly which tests are affected, and run only those tests. That is a hard problem, but my personal workflow usually doesn't need something that fancy. I usually find one test or a file of tests that I want to focus, and then make changes to the code until they pass (re-running them on every code change). If that is similar to what you need, you can try this out:

I'm using [overseer](https://github.com/stevearc/overseer.nvim) and enabling the [neotest consumer](https://github.com/stevearc/overseer.nvim/blob/master/doc/third_party.md#neotest) so that every neotest run is done as an overseer task. This unlocks a lot of new options, like custom notifications (even system notifications) when the test completes. I'll run a test with `neotest.run.run({ vim.api.nvim_buf_get_name(0) })` or similar, then I'll configure the overseer task to restart whenever I save a file (`:OverseerQuickAction watch`). Under the hood this creates a `BufWritePost` autocmd. Not as fancy as a real file watcher, but works for my purposes. Edit edit edit, work until the tests pass, then to stop it I'll just `:OverseerQuickAction dispose` to get rid of the watcher.

Bonus tip: if you have the output terminal open in a window (`:OverseerQuickAction open vsplit`), it will get updated every time the task restarts.

I should probably write up a full tutorial or do a video about this workflow, but I haven't had the time yet. Feel free to ask questions, as I kind of glossed over a lot of the details.

**Reply by rcarriga** [reply]
Yeah the problem of file watching is really complex, I've had some rough starts at it but nothing even close to concrete yet. I was looking into using the call hierarchy features of LSP to do it but they seem to be very finicky about where the location is in order to get information so it's hard to do consistently across languages/adapters

**Reply by gugahoi** [reply]
Thank you @stevearc for sharing that workflow, looks like an improvement to what I currently do. It seems the core of how we operate is the same to be honest. My workflow is to have a terminal open with the native "watcher" mode of the testing framework running which reruns on save as well.

@rcarriga Would it be feasible to try to simply leverage a framework's built in watcher somehow? For example: node/jest would simply use `--watch` when starting the test. I might be way off here tho, I haven't really looked at how neotest works under the hood.

**Reply by rcarriga** [reply]
Yes that should actually work just fine, but it'll require the adapter to support streaming results. I don't believe neotest-jest does at the moment support that because it relies on the results being dumped in a file. I'd recommend opening an issue with the Jest adapter to see if they have an idea on how feasible it would be. neotest-python achieves the streaming by integrating with pytest which might be the only way to do it for Jest as well

**Reply by jfpedroza** [reply]
It would be really nice to add that. There is a package in Elixir that allows you to watch and run all tests, failed tests, stale tests or tests that match a substring (not just all in a folder) and I have been thinking on how to integrate it with Neotest.

Let the command be in charge of deciding when to run tests, streaming them to Neotest whenever it wants.

**Reply by rcarriga** [reply]
@jfpedroza This should already be possible since the elixir adapter supports streaming, have you tried it yet?

**Reply by jfpedroza** [reply]
I actually hadn't tried it. I just tried it, and it didn't work, but it is an issue with the package and how I got the adapter to work. I'll see how to get it to work.

**Reply by jfpedroza** [reply]
I got it to work 😄.  The only thing missing is a way to reset the tests back to "running" on save.

**Reply by rcarriga** [reply]
Brilliant! Is that just an aesthetic issue? As in, everything is running and results update correctly, the user just isn't shown that it is happening?

**Reply by jfpedroza** [reply]
Yes, it's only aesthetic. You save and when new results arrive, the sign is updated. But the user doesn't know tests are running again.

**Reply by rcarriga** [reply]
I've created an issue for this so I can keep track of it :+1: https://github.com/nvim-neotest/neotest/issues/169

**Comment by Issafalcon** [comment]
Hi!

So this might be quite specific to dotnet, but there is an issue that the users are facing at the moment (including myself) in that the way dotnet test commands work. It either looks for the nearest `.csproj` file in the directory specified in the command, or the nearest `.sln` file. 

For the most part, running tests are currently done at the project level, and this works well. However, the dotnet test runner will overwrite the test output file multiple times, if it is run from the `.sln` root folder, because each projects tests are run independently, with results collected independently. The end result is that the last project to run it's tests will be the one with its results in the output file.  See the result output when I run the whole test suite from a `.sln` root folder:
![image](https://user-images.githubusercontent.com/19861614/215324441-852448e9-a9ca-497d-8536-885615313b6e.png)

My first thought is that the solution to this would be if the `adapter.build_spec` function, could return an array of specs like:

```
return {
  {
    command = command_string_1,
    context = {
      results_path = results_path_project_1,
      file = position.path_of_project_1,
      id = position.id_of_project_1,
    },
  },
 {
    command = command_string_2,
    context = {
      results_path = results_path_project_2,
      file = position.path_of_project_2,
      id = position.id_of_project_2,
    },
  }
```
}

Then they could maybe run sequentially or in parallel, each with their own process and result file (like the case would be if I hit 'run' on each of the individual projects from the summary window in quick succession).

Forgive me if you can already do something like this, or there are ways that this can be done in the adapters! If either is the case, it would be great to see how to do it so I can add this feature to the dotnet adapter. Even if there is a away to recursively call this build_spec function in the adapter, as if each project was running tests in isolation.

**Reply by rcarriga** [reply]
You can actually already return an array of specs :smile:  
https://github.com/nvim-neotest/neotest/blob/392808a91d6ee28d27cbfb93c9fd9781759b5d00/lua/neotest/adapters/interface.lua#L34 

For reference here is the original reason for the feature https://github.com/nvim-neotest/neotest/issues/93#issuecomment-1232623604

**Reply by Issafalcon** [reply]
Aha! Fantastic! Should have checked that out first 😄
